python/ray/experimental/channel/nccl_group.py

```diff
@@ -172,7 +189,7 @@ def recv(
         dtype: "torch.dtype",
         peer_rank: int,
         allocator=Optional[TorchTensorAllocator],
-    ) -> "torch.Tensor":
+    ) -> Union["torch.Tensor", Tuple["torch.Tensor", "cp.cuda.Event"]]:
```

Comment:
This API needs to be refined:
1) Consider introducing a new `recv_gpu_async()` API, but this complicates client side code which needs to decide which method to call;
2) We should not have cuda.Event in the API, may need our wrapper API for event

Reply:
Can we instead always return `Tuple["torch.Tensor", Optional["cp.cuda.Event"]]`? I think that will be a cleaner interface. You could also wrap this in our own dataclass like `MaybeAsyncTensor`.

Reply:
what about generalizing all send/recv to Future? And if it is blocking, the returned future is already ready. (I think code complexity wise it is same as returning None bcause you need to check if the second output is None anyway)

```
fut = recv()
val = fut.wait()
```

Reply:
Yes, exactly!

Reply:
Using Future sounds like a good idea. Will update.

Reply:
Done.